### ISSUE TITLE
INTERLIS Import test-dataset-DSS.xtf

### ISSUE THREAD

**Issue by sjib** [teksi/wastewater#181]
I do not understand why this error happens - there is only one dataset Mittelaletsch3 in catchment_areas :

![import_test-dataset-DSS_catchment](https://github.com/teksi/wastewater/assets/1265865/b0d40a00-ec50-4406-aaf6-30390acaff00)

`Details: (psycopg2.errors.UniqueViolation) FEHLER:  doppelter Schlüsselwert verletzt Unique-Constraint »in_od_catchment_area_identifier«
DETAIL:  Schlüssel »(identifier, fk_dataowner)=(Mittelaletsch3, ch080qwzPR000018)« existiert bereits.

[SQL: INSERT INTO tww_od.catchment_area (obj_id, direct_discharge_current, direct_discharge_planned, discharge_coefficient_rw_current, discharge_coefficient_rw_planned, discharge_coefficient_ww_current, discharge_coefficient_ww_planned, drainage_system_current, drainage_system_planned, identifier, infiltration_current, infiltration_planned, perimeter_geometry, population_density_current, population_density_planned, remark, retention_current, retention_planned, runoff_limit_current, runoff_limit_planned, seal_factor_rw_current, seal_factor_rw_planned, seal_factor_ww_current, seal_factor_ww_planned, sewer_infiltration_water_production_current, sewer_infiltration_water_production_planned, surface_area, waste_water_production_current, waste_water_production_planned, last_modification, fk_dataowner, fk_provider, fk_wastewater_networkelement_rw_current, fk_wastewater_networkelement_rw_planned, fk_wastewater_networkelement_ww_planned, fk_wastewater_networkelement_ww_current, fk_special_building_rw_planned, fk_special_building_rw_current, fk_special_building_ww_planned, fk_special_building_ww_current) VALUES (%(obj_id)s, %(direct_discharge_current)s, %(direct_discharge_planned)s, %(discharge_coefficient_rw_current)s, %(discharge_coefficient_rw_planned)s, %(discharge_coefficient_ww_current)s, %(discharge_coefficient_ww_planned)s, %(drainage_system_current)s, %(drainage_system_planned)s, %(identifier)s, %(infiltration_current)s, %(infiltration_planned)s, ST_GeomFromEWKT(%(perimeter_geometry)s), %(population_density_current)s, %(population_density_planned)s, %(remark)s, %(retention_current)s, %(retention_planned)s, %(runoff_limit_current)s, %(runoff_limit_planned)s, %(seal_factor_rw_current)s, %(seal_factor_rw_planned)s, %(seal_factor_ww_current)s, %(seal_factor_ww_planned)s, %(sewer_infiltration_water_production_current)s, %(sewer_infiltration_water_production_planned)s, %(surface_area)s, %(waste_water_production_current)s, %(waste_water_production_planned)s, %(last_modification)s, %(fk_dataowner)s, %(fk_provider)s, %(fk_wastewater_networkelement_rw_current)s, %(fk_wastewater_networkelement_rw_planned)s, %(fk_wastewater_networkelement_ww_planned)s, %(fk_wastewater_networkelement_ww_current)s, %(fk_special_building_rw_planned)s, %(fk_special_building_rw_current)s, %(fk_special_building_ww_planned)s, %(fk_special_building_ww_current)s)]
[parameters: ({'obj_id': 'ch080qwzOP001011', 'direct_discharge_current': None, 'direct_discharge_planned': None, 'discharge_coefficient_rw_current': None, 'discharge_coefficient_rw_planned': None, 'discharge_coefficient_ww_current': None, 'discharge_coefficient_ww_planned': None, 'drainage_system_current': None, 'drainage_system_planned': None, 'identifier': 'Mittelaletsch3', 'infiltration_current': None, 'infiltration_planned': None, 'perimeter_geometry': None, 'population_density_current': None, 'population_density_planned': None, 'remark': 'Mittelaletsch3', 'retention_current': None, 'retention_planned': None, 'runoff_limit_current': None, 'runoff_limit_planned': None, 'seal_factor_rw_current': None, 'seal_factor_rw_planned': None, 'seal_factor_ww_current': None, 'seal_factor_ww_planned': None, 'sewer_infiltration_water_production_current': None, 'sewer_infiltration_water_production_planned': None, 'surface_area': None, 'waste_water_production_current': None, 'waste_water_production_planned': None, 'last_modification': datetime.date(2019, 10, 14), 'fk_dataowner': 'ch080qwzPR000018', 'fk_provider': 'ch080qwzPR000017', 'fk_wastewater_networkelement_rw_current': None, 'fk_wastewater_networkelement_rw_planned': None, 'fk_wastewater_networkelement_ww_planned': None, 'fk_wastewater_networkelement_ww_current': None, 'fk_special_building_rw_planned': None, 'fk_special_building_rw_current': None, 'fk_special_building_ww_planned': None, 'fk_special_building_ww_current': None}, {'obj_id': 'ch080qwzEZ001008', 'direct_discharge_current': 5458, 'direct_discharge_planned': 5459, 'discharge_coefficient_rw_current': None, 'discharge_coefficient_rw_planned': None, 'discharge_coefficient_ww_current': None, 'discharge_coefficient_ww_planned': None, 'drainage_system_current': 5188, 'drainage_system_planned': 5193, 'identifier': 'Mittelaletsch1', 'infiltration_current': 5453, 'infiltration_planned': 5462, 'perimeter_geometry': '010a000020080800000100000001090000000100000001020000000800000000000000153144419a999919c0743141fca9f1f2e6304441713d0af74f743141c74b37e99a304441713d0a3 ... (20 characters truncated) ... 04441713d0ab763743141f853e3d5513044414260e510e6743141c74b370991304441ac1c5a641b753141fca9f1d2f03044410e2db29d0975314100000000153144419a999919c0743141', 'population_density_current': None, 'population_density_planned': None, 'remark': None, 'retention_current': 5468, 'retention_planned': 5470, 'runoff_limit_current': None, 'runoff_limit_planned': None, 'seal_factor_rw_current': None, 'seal_factor_rw_planned': None, 'seal_factor_ww_current': None, 'seal_factor_ww_planned': None, 'sewer_infiltration_water_production_current': None, 'sewer_infiltration_water_production_planned': None, 'surface_area': Decimal('99999.9000'), 'waste_water_production_current': Decimal('1.330'), 'waste_water_production_planned': Decimal('1.450'), 'last_modification': datetime.date(2019, 10, 14), 'fk_dataowner': 'ch080qwzPR000018', 'fk_provider': 'ch080qwzPR000017', 'fk_wastewater_networkelement_rw_current': None, 'fk_wastewater_networkelement_rw_planned': None, 'fk_wastewater_networkelement_ww_planned': None, 'fk_wastewater_networkelement_ww_current': None, 'fk_special_building_rw_planned': None, 'fk_special_building_rw_current': None, 'fk_special_building_ww_planned': None, 'fk_special_building_ww_current': None}, {'obj_id': 'ch080qwzEZ001009', 'direct_discharge_current': None, 'direct_discharge_planned': None, 'discharge_coefficient_rw_current': None, 'discharge_coefficient_rw_planned': None, 'discharge_coefficient_ww_current': None, 'discharge_coefficient_ww_planned': None, 'drainage_system_current': 5187, 'drainage_system_planned': None, 'identifier': 'Mittelaletsch2', 'infiltration_current': None, 'infiltration_planned': None, 'perimeter_geometry': '010a0000200808000001000000010900000001000000010200000007000000fca9f1f2e6304441713d0af74f743141333333436c3144413d0ad7c3107431416210589902314441d34d62f064733141c976be0fde304441cff75343fc72314196438b5cc6304441a01a2f5d927331416abc74d39f304441d7a3705d3f743141fca9f1f2e6304441713d0af74f743141', 'population_density_current': None, 'population_density_planned': None, 'remark': None, 'retention_current': None, 'retention_planned': None, 'runoff_limit_current': None, 'runoff_limit_planned': None, 'seal_factor_rw_current': None, 'seal_factor_rw_planned': None, 'seal_factor_ww_current': Decimal('95.00'), 'seal_factor_ww_planned': None, 'sewer_infiltration_water_production_current': Decimal('1.300'), 'sewer_infiltration_water_production_planned': None, 'surface_area': Decimal('99999.9000'), 'waste_water_production_current': Decimal('12.300'), 'waste_water_production_planned': None, 'last_modification': datetime.date(2019, 10, 14), 'fk_dataowner': 'ch080qwzPR000018', 'fk_provider': 'ch080qwzPR000017', 'fk_wastewater_networkelement_rw_current': None, 'fk_wastewater_networkelement_rw_planned': None, 'fk_wastewater_networkelement_ww_planned': None, 'fk_wastewater_networkelement_ww_current': None, 'fk_special_building_rw_planned': None, 'fk_special_building_rw_current': None, 'fk_special_building_ww_planned': None, 'fk_special_building_ww_current': None}, {'obj_id': 'ch080qwzEZ001010', 'direct_discharge_current': None, 'direct_discharge_planned': None, 'discharge_coefficient_rw_current': None, 'discharge_coefficient_rw_planned': None, 'discharge_coefficient_ww_current': None, 'discharge_coefficient_ww_planned': None, 'drainage_system_current': 5187, 'drainage_system_planned': None, 'identifier': 'Mittelaletsch3', 'infiltration_current': None, 'infiltration_planned': None, 'perimeter_geometry': '010a000020080800000100000001090000000100000001020000000a000000000000700f314441448b6c67cb7431413108acac4b314441759318a407753141000000d08f3144417593180 ... (84 characters truncated) ... 144413d0ad7c310743141e9263118673144414260e5b015743141508d977e05314441931804568d74314100000000153144419a999919c0743141000000700f314441448b6c67cb743141', 'population_density_current': None, 'population_density_planned': None, 'remark': None, 'retention_current': None, 'retention_planned': None, 'runoff_limit_current': None, 'runoff_limit_planned': None, 'seal_factor_rw_current': Decimal('50.00'), 'seal_factor_rw_planned': None, 'seal_factor_ww_current': None, 'seal_factor_ww_planned': None, 'sewer_infiltration_water_production_current': None, 'sewer_infiltration_water_production_planned': None, 'surface_area': Decimal('99999.9000'), 'waste_water_production_current': None, 'waste_water_production_planned': None, 'last_modification': datetime.date(2019, 10, 14), 'fk_dataowner': 'ch080qwzPR000018', 'fk_provider': 'ch080qwzPR000017', 'fk_wastewater_networkelement_rw_current': None, 'fk_wastewater_networkelement_rw_planned': None, 'fk_wastewater_networkelement_ww_planned': None, 'fk_wastewater_networkelement_ww_current': None, 'fk_special_building_rw_planned': None, 'fk_special_building_rw_current': None, 'fk_special_building_ww_planned': None, 'fk_special_building_ww_current': None}, {'obj_id': 'ch080qwzEZ001011', 'direct_discharge_current': None, 'direct_discharge_planned': None, 'discharge_coefficient_rw_current': None, 'discharge_coefficient_rw_planned': Decimal('67.00'), 'discharge_coefficient_ww_current': None, 'discharge_coefficient_ww_planned': Decimal('67.00'), 'drainage_system_current': 5186, 'drainage_system_planned': None, 'identifier': 'Mittelaletsch4', 'infiltration_current': None, 'infiltration_planned': None, 'perimeter_geometry': '010a0000200808000001000000010900000001000000010200000005000000508d977e05314441931804568d743141e9263118673144414260e5b0157431412b8716d9fc304441bc7493f87a743141fca9f1f2e6304441713d0af74f743141508d977e05314441931804568d743141', 'population_density_current': None, 'population_density_planned': None, 'remark': None, 'retention_current': None, 'retention_planned': None, 'runoff_limit_current': None, 'runoff_limit_planned': None, 'seal_factor_rw_current': None, 'seal_factor_rw_planned': None, 'seal_factor_ww_current': None, 'seal_factor_ww_planned': None, 'sewer_infiltration_water_production_current': None, 'sewer_infiltration_water_production_planned': None, 'surface_area': Decimal('99999.9000'), 'waste_water_production_current': None, 'waste_water_production_planned': None, 'last_modification': datetime.date(2019, 10, 14), 'fk_dataowner': 'ch080qwzPR000018', 'fk_provider': 'ch080qwzPR000017', 'fk_wastewater_networkelement_rw_current': None, 'fk_wastewater_networkelement_rw_planned': None, 'fk_wastewater_networkelement_ww_planned': None, 'fk_wastewater_networkelement_ww_current': None, 'fk_special_building_rw_planned': None, 'fk_special_building_rw_current': None, 'fk_special_building_ww_planned': None, 'fk_special_building_ww_current': None})]
(Background on this error at: http://sqlalche.me/e/13/gkpj)`

@domi4484 
For me it is strange that in class surface_runoff_parameters.identifier "Mittelaletsch3" had to be changed to Mittelaletsch5
https://github.com/teksi/wastewater/pull/111/files#diff-d3fc0d84e6c97984de7c9dd5f81443f439e385a17fbf13d9de1b1e6104bce94eR10793

If I look at the the class diagramm then class surface_runoff_parameters has an association to catchment_area and is not a subclass.
Or is there something wrong in the datamodel definition?

With the corrected version I was able to successfully import.

**Comment by sjib** [comment]
![import_test-dataset_succes](https://github.com/teksi/wastewater/assets/1265865/d7ff7170-729d-459d-91ad-ddde24bae955)

**Comment by domi4484** [comment]
@sjib I think I modified that because of that composite key constraint error. I did this in a error search session and then forgot to come back to you to ask if that constraint really make sense. 
I can imagine is quite common for different structure type to have same owner and identifier?

**Comment by urskaufmann** [comment]
I installed this evening from #72  the datamodel, plugin and project with QGIS 3.34.3
Installation is ok. When starting project the first time, the network-views are missing.
Then import the musterdatensatz (test-dataset-organisations and test-dataset-DSS). OK, no error.
Then SQL-Button: can not initialize transaction. -> Project/Properties/datasources: shows automatic transaction groups and option  Vorgabewerte auf Anbieterseite auswerten is set. I toogle these two options several times, but always this red error.
After restart of QGIS, the datasources-options are the same as initially, but now it works!! Confusing, not logical...

**Comment by sjib** [comment]
@domi4484 I think I found the problem:
class surface_runoffparameter (Oberflaechenabflussparameter) is configured as subclass of catchment_area in the matching

https://github.com/teksi/wastewater/blob/c49a468cd0232b92a4c1476932f165b61e408242/plugin/teksi_wastewater/interlis/interlis_model_mapping/model_tww_od.py#L222

instead of a 1:1 relation as described in INTERLIS:

```
!!@ comment = "Assoziation"
ASSOCIATION Oberflaechenabflussparameter_EinzugsgebietAssoc =
  EinzugsgebietRef  -- {1} Einzugsgebiet;
  Oberflaechenabflussparameter_EinzugsgebietAssocRef -- {0..1} Oberflaechenabflussparameter;
END Oberflaechenabflussparameter_EinzugsgebietAssoc;

```
![20240412_relation_surface_runoffparameter_catchment_area](https://github.com/teksi/wastewater/assets/1265865/96b603a8-c7f3-447c-940e-279f47ad5847)

It should be 
      `  class surface_runoff_parameters(self.Base):`

This would also explain why we suddenly had a duplicate identifier before it was changed to Mittelaletsch5:
![20240412_data_catchment_area_with_data_surface_area](https://github.com/teksi/wastewater/assets/1265865/125fee09-f897-4485-828e-f0d9aa1c9a14)

**Comment by sjib** [comment]
In qgep the configuration is correct

https://github.com/QGEP/qgepqwat2ili/blob/2455a6a71da44ea58e2a98b421dc35db9e657830/qgepqwat2ili/qgepdss/model_qgep.py#L397
class surface_runoff_parameters(Base):

**Comment by domi4484** [comment]
@sjib yes it is fixed in pr #191

**Comment by sjib** [comment]
Solved with https://github.com/teksi/wastewater/pull/191
